Run tests with tcpdump in CI

arnout opened the pull request:
Enable the tcpdump option of test_flows.py when running in CI.
This makes is easier to debug issues in CI, e.g. #1029.

This requires a fix taken from #1031 to make it possible to start tcpdump before test_gw_repeater.sh is called.
Cherrypicking just that single patch was a bit tricky, so I've taken along two other patches as well.

arnout committed:
test_flows.py: move opts to a separate module

When we will split up the test functionality over several classes or
files, there are some options that we need to access from multiple
places. To make access to these easier, move them to a static class
(i.e. a class that is not instantiated but is only used through class
variables) in a separate module.

Move the different message functions to this module as well. Since
that means we loose access to the currently running test, remove that
information from message(), and add it explicitly to the start and end
of test. Since ok() is used in only one place, remove it entirely. It
would be possible to keep access to the currently running test by adding
it to opts, but that is pretty ugly and not worth it - especially since
unittest and boardfarm don't have such a feature anyway.

Note that the opts class will be removed again when this is refactored
with unittest or boardfarm, because those options will then be taken
from the unittest or boardfarm environment. The message functions may
also be refactored to whatever is provided by unittest or boardfarm at
that time.

Add the opts module to the flake8 checks.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: run_tests: take tests as an argument

Instead of taking the list of tests from self.opts, take it as an
argument to run_tests. This will allow us to remove self.opts.

While we're at it, also fill in the default of all tests in run_tests
itself.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: fix tcpdump when network doesn't exist

The tcpdump option uses 'docker network inspect' to discover the name of
the bridge corresponding to the prplMesh-net-uniqueid network. However,
when tcpdump is started before the containers are started in the init()
function, the network may not exist yet, which leads to an exception.

Make sure the network does exist by creating it. This duplicates some
code from tools/docker/run.sh, but so be it.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
.gitlab-ci.yml: run tests with tcpdump

This makes it much easier to debug failures in CI.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on tests/test_flows.py:
```diff
@@ -81,8 +81,21 @@ def tcpdump_start(self):
             debug("Starting tcpdump, output file {}".format(outputfile))
-            inspect = json.loads(subprocess.check_output(('docker', 'network', 'inspect',
-                                                          'prplMesh-net-{}'.format(self.opts.unique_id))))
+            docker_network = 'prplMesh-net-{}'.format(self.opts.unique_id)
```
It will be more appropriate to move this to a separate method.

arnout replied:
Good point, but it will make rebase difficult so I'll do that later.

arnout replied:
Actually, I needed to make some other changes anyway that will lead to conflicts during the rebase, so I did it now after all.

tomereli commented on tests/test_flows.py:
```diff
@@ -81,8 +81,21 @@ def tcpdump_start(self):
+                # Raise an exception if it fails.
+                subprocess.run(('docker', 'network', 'create', docker_network), check=True,
+                               stdout=subprocess.DEVNULL)
+                # Inspect again, now raise if it fails.
```
I might be missing something here, but where is the `raise Exception` ?

arnout replied:
It's in the `check = True`.

tomereli commented on tests/test_flows.py:
```diff
@@ -81,8 +81,21 @@ def tcpdump_start(self):
+                                                stdout=subprocess.PIPE)
+
+            inspect = json.loads(inspect_result.stdout)
             prplmesh_net = inspect[0]
```
This as well should be now a separate function IMO.

tomereli commented on .gitlab-ci.yml:
```diff
@@ -101,7 +101,7 @@ run-tests:
     # Try to pull from dockerhub, if it fails, the run should fail
     - tools/docker/image-pull.sh
-    - tests/test_flows.py -v -u $RANDOM
+    - tests/test_flows.py -v -u $RANDOM --tcpdump
```
Can we make tcpdump enabled by default? Any reason not to?

arnout replied:
Well, it will be once #1020 matures.

tomereli reviewed (approved):
Minor comments but looking good so approving

arnout committed:
test_flows.py: create logs directory before starting tcpdump

If the logs directory doesn't exist yet, tcpdump will fail. Therefore,
simply always create the logs directory before starting tcpdump.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
